backends/ebpf/ebpfTable.cpp

```diff
@@ -151,54 +186,70 @@ void EBPFTable::emitActionArguments(CodeBuilder* builder,
 }
 
 void EBPFTable::emitValueType(CodeBuilder* builder) {
-    // create an enum with tags for all actions
+    emitValueActionIDNames(builder);
+
+    // a type-safe union: a struct with a tag and an union
     builder->emitIndent();
-    builder->append("enum ");
-    builder->append(actionEnumName);
-    builder->spc();
+    builder->appendFormat("struct %s ", valueTypeName.c_str());
     builder->blockStart();
 
+    emitValueStructStructure(builder);
+
+    builder->blockEnd(false);
+    builder->endOfStatement(true);
+}
+
+void EBPFTable::emitValueActionIDNames(CodeBuilder* builder) {
+    // create type definition for action
+    builder->emitIndent();
+    unsigned int action_idx = 1;  // 0 is reserved for NoAction
     for (auto a : actionList->actionList) {
         auto adecl = program->refMap->getDeclaration(a->getPath(), true);
         auto action = adecl->getNode()->to<IR::P4Action>();
-        cstring name = EBPFObject::externalName(action);
+        // NoAction is from standard library
+        if (action->name.originalName == P4::P4CoreLibrary::instance.noAction.name) {
```

Comment:
so why not define that as 0?

Reply:
We don't need to define a constant for `NoAction`, because we explicitly generate `case 0:` here: https://github.com/p4lang/p4c/pull/3129/files#diff-218b38cacc617dd2223b207e565c9b04ace178cfc080f199e48832a3266f3ebdR439.